### ISSUE TITLE
Migrating extensions from Brackets

### ISSUE THREAD

**Issue by michaljerabek** [phcode-dev/phoenix-desktop#380]
Hi, 

I have two extensions I cannot make to work in Phoenix:

1. Unpublished one, where the main idea is to run [Cheerio](https://github.com/cheeriojs/cheerio/) on a selected HTML, but it can run any code on any selection.
    - The code is executed in Node, but `NodeDomain` module in Phoenix is not connecting to Node. I've noticed, there is `NodeConnector` now, but it doesn't seem to be usable inside an extension.
2. [Crown Control](https://github.com/michaljerabek/Crown-Control) for integrating Logitech Craft Keyboard (the dial control)
    - To make it work I need to get PID (process id) of Phoenix. In Brackets it was also get by `NodeDomain`, which is nonsens, because it returns PID of Node, but it actually works with any existing PID. Is it possible to get the PID? (Or make it possible?)
    - But even if I instert the PID manually, I need to make WebSocket connection to `ws://127.0.0.01:10134` to connect to Logitech Options. Here I get the following error: `failed: Error during WebSocket handshake: Unexpected response code: 403`. I guess there is some security policy and the address needs to be allowed.

Thank you for any help.

**Comment by abose** [comment]
Hi @michaljerabek , thank you for contributing to extensions. Most extensions till now were targeted to browsers as the new desktop builds are only about a month old. So we decided to defer exposing NodeConnector till an extension author requested for it. I will try to get it in for the May release of phcode. But, you can test it out as it becomes available through the dev builds even before that. 

As pointed out, `NodeConnector` is our new node API . Please review if this API spec works for you [NodeConnector API spec](https://github.com/phcode-dev/phoenix/wiki/NodeConnector-API).

Phoenix have a bit more stricter security posture than Brackets as the plan is to gradually increase sandboxing to be able to run un-trusted extensions in a safe way. That is one of the reasons why we had to deprecate the old `NodeDomain`. Also the old Domain code API was not pretty to work with, was built on an ancient version of node which we could not port and had a lot of compatibility issues. So we decided to ditch that in favor of modern node-20 with built in sandboxing.

**Comment by abose** [comment]
Reg. extension 2, `Crown Control`

1. I guess it will be better to add a shell API- `Phoenix.app.getProcessID()` . will add that.
2. Due to the new security posture, ws is only allowed from `ws://localhost:` , so replacing `12.0.0.1* with `localhost` should fix it. I will add `ws://127.0.0.1` to the list too.

Is there any other changes you need to get things working?

**Comment by michaljerabek** [comment]
Thank you for your replay.

If there will be a way to load a file from extension in Node, `NodeConnector` should be working fine. (I think.)

If I use `localhost` it throws the same error. Maybe the issue is something else. I'll try to research this more.

**Comment by michaljerabek** [comment]
The error is just: `WebSocket connection to 'ws://localhost:10134/' failed: Error during WebSocket handshake: Unexpected response code: 403`. There is no more information.
Thanks!

**Comment by abose** [comment]
> The error is just: `WebSocket connection to 'ws://localhost:10134/' failed: Error during WebSocket handshake: Unexpected response code: 403`. There is no more information. Thanks!

The HTTP 403 Forbidden response status code indicates that the server understands the request but refuses to authorize it. The connection is going through from phcode, but the server is not accepting it as authorized.

**Comment by abose** [comment]
@michaljerabek it appears that exposing node API to extensions is a bit more harder than i anticipated. Are you looking at any particular time frame for this?

**Comment by michaljerabek** [comment]
No, I am not. Do it well rather than quickly. This level of service is much more than I was expecting anyway. :)

**Comment by michaljerabek** [comment]
> > The error is just: `WebSocket connection to 'ws://localhost:10134/' failed: Error during WebSocket handshake: Unexpected response code: 403`. There is no more information. Thanks!
> 
> The HTTP 403 Forbidden response status code indicates that the server understands the request but refuses to authorize it. The connection is going through from phcode, but the server is not accepting it as authorized. Does the server need any auth headers?

It shouldn't need anything. Everywere I've been looking it's implemented as `new WebSockect("ws://127.0.0.1:10134")`.

**Comment by abose** [comment]
It should be pretty straightforward to do websockets from node i think. 

BTW, in phcode, you can directly load the current project as extension unlike in brackets, if you haven't found it.
![image](https://github.com/phcode-dev/phoenix-desktop/assets/5336369/b7f1ba37-82bf-4de9-ba45-14d174cb4fe2)

> No, I am not. Do it well rather than quickly. This level of services is much more than I was expecting anyway. :)

I have completed the extension loading code, but the main problem is that i have to update the backend services to validate node extensions, write docs and integ tests. But i can share the build with node extension load support(already done) in a day or two so that you are unblocked. Full rollout of the feature would only happen by May/June monthly release.

## Basic node extension file organisation
The following change is needed in `package.json` for enabling node
```js
"nodeConfig": {
        "nodeIsRequired": true, // set to false if you want to run the extension in browser as well.
        "main": "node/index.js",
        "npmInstall": "node/" // this is optional, if given, it will run npm install on the given extension folder after installing. this is to bring in external libraries. 
    }
```
### `main.js` - main.js loaded in brackets
```js
define(function (require, exports, module) {
    const NodeConnector = brackets.getModule("NodeConnector");
    const AppInit = brackets.getModule("utils/AppInit");
    const nodeConnector = NodeConnector.createNodeConnector("extension_connector_1", exports);
    
    AppInit.appReady(function () {
        nodeConnector.execPeer("echoTest").then(alert);
    });
});
```

### `node/index.js` - node main file
```js
const extnNodeConnector = global.createNodeConnector("extension_connector_1", exports);

async function echoTest(){
    return "hello from node";
}

exports.echoTest = echoTest;
```

Does this APIs look eazy to work with? open to suggestions.

**Comment by michaljerabek** [comment]
Thanks, it's looking good!

**Comment by abose** [comment]
@ Dev builds with node extension support is now available at https://github.com/phcode-dev/phoenix-desktop/releases/tag/dev-app-v3.7.2

To create a node js extension, use this template: https://github.com/phcode-dev/extension-node-template

Full support should be available in installed builds starting with may end release. Please provide any feedback/improvements.

**Comment by abose** [comment]
Closing as done. Please reopen this issue or create new issues as needed for anything.
